File: runtime/cpp/tests/peer.cpp
#include <nightseam/runtime/peer.hpp>
#include <nightseam/duplex/ws.hpp>
#include <future>
#include <iostream>
#include <mutex>
#include <thread>
using namespace nightseam::runtime;
using namespace nightseam::duplex;
using namespace std::chrono_literals;
static void require(bool value, const char* message) { if (!value) throw std::runtime_error(message); }
static void websocket_close() {
    ws::Listener listener;
    auto dialled=ws::dial(listener.url());
    auto accepted=listener.accept(Wait::after(2s));
    PeerOptions options; options.write_timeout=1s;
    Peer server(accepted.conn,Role::server,options), client(dialled.conn,Role::client,options);
    server.handle("echo",[](const RequestContext&,Peer&,const Value& value){return value;});
    require(client.call("echo",Value(1))==Value(1),"WebSocket peer exchange");
    client.close();
    auto local=client.await_close(Wait::after(2s)), remote=server.await_close(Wait::after(2s));
    require(local.clean && remote.clean && local.code==1000 && remote.code==1000,"WebSocket chosen close reaches both peers");
}

static void raw_payload_and_metadata() {
    auto [raw,transport]=pipe();
    PeerOptions options;
    options.prepare=[](Peer& peer) {
        peer.handle_raw("echo",[](const RequestContext& ctx,Peer&,const Value&) {
            require(ctx.meta && ctx.meta->at("tenant")=="secret","metadata reaches handler");
            return ctx.raw_payload;
        });
    };
    Peer peer(transport,Role::server,options);
    std::string payload=R"({ "n":1e3, "duplicate":1, "duplicate":2, "nil":null })";
    raw->send({Kind::text,"{\"version\":1,\"kind\":\"request\",\"id\":\"c:9\",\"method\":\"echo\",\"params\":"+payload+",\"meta\":{\"tenant\":\"secret\"}}"});
    auto reply=decode_envelope(raw->receive(Wait::after(2s)).data,Role::client);
    require(reply.raw_result==payload,"raw handler preserves JSON value bytes");
    require(!reply.meta,"response does not carry metadata");
    peer.close();
}

static void pending_limit_and_observer() {
    auto [raw,transport]=pipe();
    std::mutex mutex; std::vector<Observation> seen;
    PeerOptions options; options.max_pending_requests=1;
    options.observer=[&](const Observation& event){std::lock_guard lock(mutex);seen.push_back(event);};
    Peer peer(transport,Role::client,options);
    std::stop_source stop;
    auto call=std::async(std::launch::async,[&]{CallOptions o; o.wait.stop=stop.get_token(); try { peer.call("wait",Value(1),o); } catch (const Cancelled&) {} });
    auto request=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
    bool busy=false;
    try { peer.call("unsent"); } catch (const PublicError& e) {busy=e.code=="busy";}
    require(busy,"pending requests are bounded locally");
    stop.request_stop(); call.get();
    auto cancel=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
    require(cancel.kind=="cancel" && cancel.id==request.id && cancel.trace.parent==request.trace.parent,"cancel correlation and trace");
    {
        std::lock_guard lock(mutex);
        std::size_t ended=seen.size(),sent=seen.size();
        for (std::size_t i=0;i<seen.size();++i) {
            require(seen[i].method!="unsent","refused call has no observer lifecycle");
            if (seen[i].type=="request.ended") { ended=i;require(seen[i].outcome=="cancelled","local cancellation outcome"); }
            if (seen[i].type=="frame.sent" && seen[i].kind=="cancel") sent=i;
        }
        require(ended<sent,"request ends before cancellation is sent");
    }
    peer.close();
}

static void cancellation_has_reserved_queue_capacity() {
    auto [raw,transport]=pipe(1<<20,1);
    PeerOptions options; options.queue_capacity=1; options.write_timeout=2s;
    Peer peer(transport,Role::client,options);
    std::stop_source stop;
    auto call=std::async(std::launch::async,[&]{CallOptions o; o.wait.stop=stop.get_token(); try { peer.call("wait",Value::null(),o); } catch (const Cancelled&) {} });
    auto request=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
    for (int i=1;i<=3;++i) peer.emit("queued",Value(i));
    stop.request_stop();
    require(call.wait_for(500ms)==std::future_status::ready,"saturated data queue does not strand cancellation"); call.get();
    for (int i=1;i<=3;++i) {
        auto frame=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
        require(frame.kind=="event" && frame.data==Value(i),"reserved control retains accepted frame order");
    }
    auto cancel=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
    require(cancel.kind=="cancel" && cancel.id==request.id,"reserved cancellation delivered"); peer.close();
}

static void inbound_stall_ends_peer() {
    auto [raw,transport]=pipe();
    std::promise<void> started; std::atomic_bool stalled=false;
    PeerOptions options; options.queue_capacity=1; options.write_timeout=30ms;
    options.observer=[&](const Observation& event){if(event.type=="backpressure" && event.stalled) stalled=true;};
    options.events["block"]=[&](const RequestContext& ctx,Peer&,const Value&){ started.set_value(); while(!ctx.cancelled()) std::this_thread::sleep_for(1ms); };
    Peer peer(transport,Role::server,options);
    auto event=R"({"version":1,"kind":"event","event":"block","data":null})";
    raw->send({Kind::text,event}); started.get_future().wait();
    raw->send({Kind::text,event}); raw->send({Kind::text,event});
    auto close=peer.await_close(Wait::after(1s)); require(!close.clean && stalled,"inbound stall is observed and ends peer");
}
int main() {
    try {
        websocket_close(); raw_payload_and_metadata(); pending_limit_and_observer();
        cancellation_has_reserved_queue_capacity(); inbound_stall_ends_peer();
        auto [a,b] = pipe();
        PeerOptions options; options.request_timeout = 2s;
        options.handlers["echo"] = [](const RequestContext&, Peer&, const Value& v) { return v; };
        options.handlers["reverse"] = [](const RequestContext& ctx, Peer& p, const Value& v) {
            CallOptions o; o.context = ctx; return p.call("echo",v,o);
        };
        options.handlers["wait"] = [](const RequestContext& ctx, Peer&, const Value&) {
            while (!ctx.cancelled()) std::this_thread::sleep_for(1ms);
            ctx.check(); return Value::null();
        };
        Peer server(b,Role::server,options), client(a,Role::client,options);
        require(client.call("reverse",parse_value("{\"optional\":null,\"empty\":[],\"number\":1e3}")) == parse_value("{\"optional\":null,\"empty\":[],\"number\":1e3}"), "reverse/presence");
        std::stop_source stop;
        CallOptions cancellation; cancellation.wait.stop = stop.get_token();
        auto call = std::async(std::launch::async,[&] {
            try { client.call("wait",Value::null(),cancellation); return false; }
            catch (const Cancelled&) { return true; }
        });
        stop.request_stop(); require(call.get(),"call cancellation");
        CallOptions timeout; timeout.timeout = 20ms;
        bool deadline = false;
        try { client.call("wait",Value::null(),timeout); } catch (const DeadlineExceeded&) { deadline = true; }
        require(deadline,"deadline is distinct from cancellation");
        require(client.call("echo",Value(42)) == Value(42),"late reply does not poison next call");
        server.identity("test",std::string(64,'a')); client.check_identity("test");
        bool mismatch = false;
        try { client.check_identity("wrong"); } catch (const PublicError& e) { mismatch = e.code == "contract_mismatch"; }
        require(mismatch,"identity mismatch");
        client.close(); require(client.await_close(Wait::after(1s)).code == 1000,"chosen close");

        // A raw transport proves a cancelled incoming request is retained until
        // its handler returns, and the response carries the original trace.
        auto [raw,transport] = pipe();
        std::promise<void> started, release;
        auto released = release.get_future().share();
        PeerOptions held;
        held.handlers["held"] = [&](const RequestContext&, Peer&, const Value&) { started.set_value(); released.wait(); return Value(7); };
        Peer peer(transport,Role::server,held);
        raw->send({Kind::text,R"({"version":1,"kind":"request","id":"c:1","method":"held","params":null,"tracestate":"verbatim"})"});
        started.get_future().wait();
        raw->send({Kind::text,R"({"version":1,"kind":"cancel","id":"c:1"})"});
        bool waiting = false;
        try { raw->receive(Wait::after(30ms)); } catch (const DeadlineExceeded&) { waiting = true; }
        require(waiting,"cancel waits for handler return"); release.set_value();
        auto response = decode_envelope(raw->receive(Wait::after(1s)).data,Role::client);
        require(response.error && response.error->code == "cancelled", "cancelled handler result");
        require(response.trace.state == "verbatim","trace echo"); peer.close();
        std::cout << "peer lifecycle invariants passed\n";
    } catch (const std::exception& error) { std::cerr << error.what() << '\n'; return 1; }
}

File: runtime/rs/tests/peer.rs
use bitwire::{Payload, PublicError};
use nightseam::{Context, Options, Peer, Role};
use nightseam_duplex::pipe;
use serde_json::json;
use std::{sync::Arc, time::Duration};
use tokio::{
    sync::{Mutex, Notify},
    time::timeout,
};

fn payload(value: serde_json::Value) -> Payload {
    Payload::from_value(&value).unwrap()
}
fn pair(options: Options) -> (Peer, Peer) {
    let (a, b) = pipe(1 << 20);
    (
        Peer::over(a, Role::Client, Options::default()).unwrap(),
        Peer::over(b, Role::Server, options).unwrap(),
    )
}

#[tokio::test]
async fn call_reverse_call_null_and_public_error() {
    let (client, server) = pair(Options::default());
    client
        .handle("reverse", |_, value| async move { Ok(value) })
        .unwrap();
    server
        .handle("echo", |_, value| async move { Ok(value) })
        .unwrap();
    server
        .handle("relay", |ctx, value| async move {
            ctx.peer()
                .unwrap()
                .call_with(ctx, "reverse", value)
                .result()
                .await
        })
        .unwrap();
    server
        .handle("deny", |_, _| async {
            Err(PublicError::new("denied", "No access"))
        })
        .unwrap();
    for method in ["echo", "relay"] {
        for value in [json!(null), json!({"text":"hello"}), json!([1, true])] {
            assert_eq!(
                client
                    .call(method, payload(value.clone()))
                    .result()
                    .await
                    .unwrap()
                    .value()
                    .unwrap(),
                value
            );
        }
    }
    assert_eq!(
        client
            .call("deny", Payload::Absent)
            .result()
            .await
            .unwrap_err()
            .code,
        "denied"
    );
    assert_eq!(
        client
            .call("missing", Payload::Absent)
            .result()
            .await
            .unwrap_err()
            .code,
        "method_not_found"
    );
    client.close();
    assert_eq!(
        timeout(Duration::from_secs(1), server.wait_closed())
            .await
            .unwrap()
            .code,
        1000
    );
}

#[tokio::test]
async fn cancellation_signals_work_but_does_not_answer_until_it_returns() {
    let (client, server) = pair(Options {
        max_concurrent_handlers: 1,
        ..Options::default()
    });
    let started = Arc::new(Notify::new());
    let cancelled = Arc::new(Notify::new());
    let release = Arc::new(Notify::new());
    server
        .handle("hold", {
            let started = started.clone();
            let cancelled = cancelled.clone();
            let release = release.clone();
            move |ctx, _| {
                let started = started.clone();
                let cancelled = cancelled.clone();
                let release = release.clone();
                async move {
                    started.notify_one();
                    ctx.cancelled().await;
                    cancelled.notify_one();
                    release.notified().await;
                    Ok(payload(json!("late")))
                }
            }
        })
        .unwrap();
    let call = client.call("hold", Payload::Absent);
    started.notified().await;
    call.cancel();
    assert_eq!(call.result().await.unwrap_err().code, "cancelled");
    cancelled.notified().await;
    assert_eq!(
        client
            .call("hold", Payload::Absent)
            .result()
            .await
            .unwrap_err()
            .code,
        "busy"
    );
    release.notify_one();
    server
        .handle("echo", |_, value| async move { Ok(value) })
        .unwrap();
    // The response releases the slot, and an unrelated call proves the late
    // cancelled response did not close or corrupt the client's correlation.
    let result = timeout(Duration::from_secs(1), async {
        loop {
            match client.call("echo", payload(json!(1))).result().await {
                Err(e) if e.code == "busy" => tokio::task::yield_now().await,
                answer => break answer,
            }
        }
    })
    .await
    .unwrap()
    .unwrap();
    assert_eq!(result.value().unwrap(), json!(1));
}

#[tokio::test]
async fn deadlines_are_local_timeouts_and_outstanding_calls_are_bounded() {
    let (a, b) = pipe(1 << 20);
    let client = Peer::over(
        a,
        Role::Client,
        Options {
            max_pending_requests: 1,
            request_timeout: Duration::from_millis(50),
            ..Options::default()
        },
    )
    .unwrap();
    let server = Peer::over(b, Role::Server, Options::default()).unwrap();
    server
        .handle("wait", |ctx, _| async move {
            ctx.cancelled().await;
            Ok(Payload::Absent)
        })
        .unwrap();
    let first = client.call("wait", Payload::Absent);
    assert_eq!(
        client
            .call("wait", Payload::Absent)
            .result()
            .await
            .unwrap_err()
            .code,
        "busy"
    );
    assert_eq!(first.result().await.unwrap_err().code, "request_timeout");
    server.close();
}

#[tokio::test]
async fn events_are_ordered_and_metadata_is_forwarded_only_explicitly() {
    let (client, server) = pair(Options::default());
    let seen = Arc::new(Mutex::new(Vec::new()));
    let notify = Arc::new(Notify::new());
    server
        .on_event("item", {
            let seen = seen.clone();
            let notify = notify.clone();
            move |ctx, value| {
                let seen = seen.clone();
                let notify = notify.clone();
                async move {
                    seen.lock()
                        .await
                        .push((value.value().unwrap(), ctx.meta().clone()));
                    notify.notify_one();
                }
            }
        })
        .unwrap();
    for n in 0..8 {
        client.emit("item", payload(json!(n))).await.unwrap();
    }
    timeout(Duration::from_secs(1), async {
        loop {
            if seen.lock().await.len() == 8 {
                break;
            }
            notify.notified().await;
        }
    })
    .await
    .unwrap();
    assert_eq!(
        seen.lock()
            .await
            .iter()
            .map(|x| x.0.clone())
            .collect::<Vec<_>>(),
        (0..8).map(|n| json!(n)).collect::<Vec<_>>()
    );
    client
        .handle("metadata", |ctx, _| async move {
            Payload::from_value(ctx.meta()).map_err(|e| PublicError::new("invalid", e))
        })
        .unwrap();
    server
        .handle("relay", |ctx, _| async move {
            assert_eq!(ctx.meta().get("tenant").map(String::as_str), Some("acme"));
            let peer = ctx.peer().unwrap();
            let inherited = peer
                .call_with(ctx.clone(), "metadata", Payload::Absent)
                .result()
                .await?;
            let explicit = peer
                .call_with(
                    ctx.clone().with_meta(ctx.meta().clone()),
                    "metadata",
                    Payload::Absent,
                )
                .result()
                .await?;
            Ok(payload(json!([
                inherited.value().unwrap(),
                explicit.value().unwrap()
            ])))
        })
        .unwrap();
    let ctx = Context::default().with_meta(
        [
            ("tenant".into(), "acme".into()),
            ("nightseam.private".into(), "dropped".into()),
        ]
        .into(),
    );
    assert_eq!(
        client
            .call_with(ctx, "relay", Payload::Absent)
            .result()
            .await
            .unwrap()
            .value()
            .unwrap(),
        json!([{}, {"tenant":"acme"}])
    );
}
